rust/lance-datagen/src/generator.rs

```diff
@@ -0,0 +1,1126 @@
+    pub fn rand<DataType>() -> Box<dyn ArrayGenerator>
+    where
+        DataType::Native: Copy + 'static,
+        PrimitiveArray<DataType>: From<Vec<DataType::Native>> + 'static,
+        DataType: ArrowPrimitiveType,
+        rand::distributions::Standard: rand::distributions::Distribution<DataType::Native>,
+    {
+        Box::new(
+            FnGen::<DataType::Native, PrimitiveArray<DataType>, _>::new_known_size(
+                DataType::DATA_TYPE.clone(),
+                move |rng| rng.gen(),
+                1,
+                DataType::DATA_TYPE
+                    .primitive_width()
+                    .map(|width| ByteCount::from(width as u64))
+                    .expect("Primitive types should have a fixed width"),
+            ),
+        )
+    }
+
+    /// Create a generator of 1d vectors (of a primitive type) consisting of randomly sampled primitive values
+    pub fn rand_vec<DataType>(dimension: Dimension) -> Box<dyn ArrayGenerator>
+    where
+        DataType::Native: Copy + 'static,
+        PrimitiveArray<DataType>: From<Vec<DataType::Native>> + 'static,
+        DataType: ArrowPrimitiveType,
+        rand::distributions::Standard: rand::distributions::Distribution<DataType::Native>,
+    {
+        let underlying = rand::<DataType>();
+        cycle_vec(underlying, dimension)
+    }
+
+    /// Create a generator of randomly sampled date32 values
+    ///
+    /// Instead of sampling the entire range, all values will be drawn from the last year as this
+    /// is a more common use pattern
+    pub fn rand_date32() -> Box<dyn ArrayGenerator> {
+        let data_type = DataType::Date32;
+        let now_ms = chrono::Utc::now().timestamp_millis();
+        let now_days = (now_ms / MS_PER_DAY) as i32;
+        let one_year_ago = now_days - 365;
+        let dist = Uniform::new(one_year_ago, now_days);
+
+        Box::new(FnGen::<i32, Date32Array, _>::new_known_size(
+            data_type.clone(),
+            move |rng| dist.sample(rng),
+            1,
+            DataType::Date32
+                .primitive_width()
+                .map(|width| ByteCount::from(width as u64))
+                .expect("Date32 should have a fixed width"),
+        ))
+    }
+
+    /// Create a generator of randomly sampled date64 values
+    ///
+    /// Instead of sampling the entire range, all values will be drawn from the last year as this
+    /// is a more common use pattern
+    pub fn rand_date64() -> Box<dyn ArrayGenerator> {
+        let data_type = DataType::Date64;
+        let now_ms = chrono::Utc::now().timestamp_millis();
+        let now_days = (now_ms / MS_PER_DAY) as i32;
+        let one_year_ago = now_days - 365;
+        let dist = Uniform::new(one_year_ago, now_days);
+
+        Box::new(FnGen::<i64, Date64Array, _>::new_known_size(
+            data_type.clone(),
+            move |rng| (dist.sample(rng) as i64) * MS_PER_DAY,
```

Comment:
It would but Arrow requires that a date64 value be exactly divisible by MS_PER_DAY:

> /// * Milliseconds (64 bits) indicating UNIX time elapsed since the epoch (no
///   leap seconds), where the values are evenly divisible by 86400000

I don't think your approach satisfies this requirement (though maybe it would be faster to modulo?  I can try it)

Reply:
Oh, yeah forgot about that requirement. Per row multiplication is likely the better option then!